### ISSUE TITLE
Neural Style in python implementation

### ISSUE THREAD

**Issue by Atraii** [apache/incubator-mxnet#2436]
I'm seeing some issues in the updated [neural style](https://github.com/dmlc/mxnet/tree/master/example/neural-style), namely:
1. The target_shape has been deprecated. Shape should be used instead. See style_gram_symbol() on line 106:<br>
   `x = mx.sym.Reshape(style[i], target_shape=(int(shape[1]), int(np.prod(shape[2:]))))` should be changed to:<br>
   `x = mx.sym.Reshape(style[i], shape=(int(shape[1]), int(np.prod(shape[2:]))))`
2.  What is grad_scale? It is using 'np.prod(shape[1:] \* shape[1]', shouldn't this be 'np.prod(shape[1:] **2)' ??? The terms should be N^2 \* M^2?
3.  Where are the constants on the style_loss? According to the paper, loss = 1 / (4 \* N^2 \* M^2) \* sum((G - A)^2). I only see the the sum of the squared difference in the get_loss function. 
4.  Is this using pre-normalized VGG-net weights? If not, shouldn't L1 normalization happen on both style and content gradients? This tends to give much better results and has been incorporated into https://github.com/jcjohnson/neural-style.
5.  Is using SGD appropriate? L-BFGS would be best to compare to the original author, but I haven't seen that implemented (another issue perhaps?) in MXNet. The second alternative would be ADAM, which is currently working.

I'm not an expert, so I would love to hear some thoughts on these.

**Comment by antinucleon** [comment]
1. Yes. I have a refactor version based on module, but not commit yet
2. It is scaling down the gram L2 loss
3. The constant doesn't matter
4. It is using VGG
5. SGD is totally fine. If you adjust the pre-defined hyper-parameters, you will get better result. eg:
![image](https://cloud.githubusercontent.com/assets/1258974/16136702/af3f85ca-33e1-11e6-871d-6aacb1b6ba7e.png)

**Comment by Atraii** [comment]
Thanks for the awesome reply!

Which hyper-parameters are you adjusting to achieve that result? Just the learning rate, style-weight, and content weight? Or are you including the momentum, etc.?

**Comment by antinucleon** [comment]
For this sample, I changed style weight and stop eps. I trained more epochs.

**Comment by Atraii** [comment]
Okay, so your commit should fix 1. and everything else is accounted for. 

Is there somewhere I can read more about how the gram L2 loss is being scaled down? I still don't fully understand what is happening with grad_scale and grad_array.

**Comment by antinucleon** [comment]
grad_scale is the normalization factor based on the gram matrix. Then we use weight and these scale to calculation make top gradient, store in grad_array. If replace that with LinearRegressionOutput, the grad_array value will be grad_scale.

**Comment by Atraii** [comment]
I'm sorry, but I must be missing something. How is the inverse of channels**2 * img_height * img_width is the gradient?

**Comment by phunterlau** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
